pkg/uploadserver/uploadserver.go

```diff
@@ -211,6 +215,96 @@ func (app *uploadServerApp) healthzHandler(w http.ResponseWriter, r *http.Reques
 	io.WriteString(w, "OK")
 }
 
+func (app *uploadServerApp) uploadHandlerAsync(w http.ResponseWriter, r *http.Request) {
+	if r.Method == "HEAD" {
+		w.WriteHeader(http.StatusOK)
+		return
+	}
+
+	if r.Method != "POST" {
+		w.WriteHeader(http.StatusNotFound)
+		return
+	}
+
+	if r.TLS != nil {
+		found := false
+
+		for _, cert := range r.TLS.PeerCertificates {
+			if cert.Subject.CommonName == app.clientName {
+				found = true
+				break
+			}
+		}
+
+		if !found {
+			w.WriteHeader(http.StatusUnauthorized)
+			return
+		}
+	} else {
+		klog.V(3).Infof("Handling HTTP connection")
+	}
+
+	exit := func() bool {
+		app.mutex.Lock()
+		defer app.mutex.Unlock()
+
+		if app.uploading || app.processing {
+			w.WriteHeader(http.StatusServiceUnavailable)
+			return true
+		}
+
+		if app.done {
+			w.WriteHeader(http.StatusConflict)
+			return true
+		}
+
+		app.uploading = true
+		return false
+	}()
+
+	if exit {
+		klog.Warning("Got concurrent upload request")
+		return
+	}
+
+	cdiContentType := r.Header.Get(UploadContentTypeHeader)
+
+	klog.Infof("Content type header is %q\n", cdiContentType)
+
+	processor, err := uploadProcessorFuncAsync(r.Body, app.destination, app.imageSize, cdiContentType)
+
+	app.mutex.Lock()
+
+	if err != nil {
+		klog.Errorf("Saving stream failed: %s", err)
+		w.WriteHeader(http.StatusInternalServerError)
+		app.uploading = false
+		app.mutex.Unlock()
+		return
+	}
+	defer app.mutex.Unlock()
+
+	app.uploading = false
+	app.processing = true
+
+	// Start processing.
+	go func() {
+		defer close(app.doneChan)
+		if err := processor.ProcessDataResume(); err != nil {
+			klog.Errorf("Error during resumed processing: %v", err)
+			// TODO: figure out how to nicely close with an error instead of just dying.
```

Comment:
if you look at the `Run` function, it is waiting on a `chan error` as well as the "done" channel.  Maybe make the error channel a struct member and use it here?  But that is assuming you want the pod to exit on error.  And I'm not sure you necessarily want that?